liferay-appsec/liferay-portal
# LPS-188676 - As an Administrator I want to be able to give ONLY activation permission for Roles (LPS-188420)

https://liferay.atlassian.net/browse/LPS-188676

Original pr - https://github.com/liferay-appsec/liferay-portal/pull/1388  by @ruben-pulido
Other pr (remove feature flag) - https://github.com/liferay-appsec/liferay-portal/pull/1390 by @ruben-pulido 
Great job Ruben!

## Changes
Splited the pr to organize them by stories.

## Original comment
Story:
https://liferay.atlassian.net/browse/LPS-188676

Tasks:
https://liferay.atlassian.net/browse/LPS-195996 (Development)
https://liferay.atlassian.net/browse/LPS-196109 (Backend tests)
Proposed Solution

This PR adds the possibility to assign two new permissions "Activate" and "Deactivate" for users.

The new permissions will be visible regardless of the feature flag, however they will only be taken into consideration if assigned if the feature flag associated to the epic is enabled (LPS-188420)

The behavior associated to the Delete permission has not changed.
A user with Delete permission can delete, activate or deactivate another user (regardless of the value of the feature flag)
Steps to Verify
Preparatory steps

    Make sure you have enabled feature flag LPS-188420
    Create several users via UI or headless API: u1, u2, u3, uA, uB
    Create several regular roles: r1, r2, r3
    For r1, assign View and Deactivate permissions for User
    For r2, assign View and Activate permissions for User
    For r3, assign View and Delete permissions for User
    Assign role r1 to user u1
    Assign role r2 to user u2
    Assign role r3 to user u3

User u1

    Log in with user u1
    Navigate to Users Admin Page
    Verify that user uA is listed under Active users
    Verify that user uA is not listed under Inactive users
    Verify that Deactivate is available for user uA
    Verify that Delete is not available for user uA
    Verify that Activate is not available for user uA
    Click on Deactivate for user uA
    Verify that user uA is no longer listed under Active users
    Verify that user uA is now listed under Inactive users
    Verify that Delete is not available for user uA
    Verify that Activate is not available for user uA
    Verify that Deactivate is not available for user uA

User u2

    Log in with user u2
    Navigate to Users Admin Page
    Verify that user uB is listed under Active users
    Verify that user uB is not listed under Inactive users
    Verify that Activate is not available for user uB
    Verify that Deactivate is not available for user uB
    Verify that Delete is not available for user uB
    Verify that user uA is not listed under Active users
    Verify that user uA is listed under Inactive users
    Verify that Delete is not available for user UA
    Verify that Activate is available for user uA
    Verify that Activate is not available for user uA
    Click on Activate for user uA
    Verify that user uA is no longer listed under Inactive users
    Verify that user uA is now listed under Active users again
    Verify that Delete is not available for user uA
    Verify that Activate is not available for user uA
    Verify that Deactivate is not available for user uA

User u3

    Log in with user u3
    Navigate to Users Admin Page
    Verify that user uB is listed under Active users
    Verify that user uB is not listed under Inactive users
    Verify that Activate is not available for user uB
    Verify that Deactivate is available for user uB
    Verify that Delete is not available for user uB
    Click on Deactivate for user uB
    Verify that user uB is no longer listed under Active users
    Verify that user uB is now listed under Inactive users again
    Verify that Delete is available for user uB
    Verify that Activate is available for user uB
    Verify that Deactivate is not available for user uB
    Click on Activate for user uB
    Verify that user uB is no longer listed under Inactive users
    Verify that user uB is now listed under Active users again
    Verify that Delete is not available for user uB
    Verify that Activate is not available for user uB
    Verify that Deactivate is available again for user uB
    Click on Deactivate for user uB
    Verify that user uB is no longer listed under Active users
    Verify that user uB is now listed under Inactive users again
    Verify that Delete is available for user uB
    Verify that Activate is available for user uB
    Verify that Deactivate is not available for user uB
    Click on Delete for user uB
    Verify that user uB is no longer listed under Active users
    Verify that user uB is no longer listed under Inactive users
